File: services/web/src/domain/tickets.js
// Ticket validity mirror of the server engine (build prompt Table 4) so an
// offline client can show a ticket's status without a round trip:
//
//   expiry = entry_timestamp + (nights_purchased x 24h)
//   status = "Active" if now < expiry else "Expired"
//
// The server remains authoritative; this is display-only. All arithmetic is
// in UTC milliseconds so a device's local timezone cannot skew the result.

export const HOURS = 60 * 60 * 1000;

export function computeExpiry(entryTimestamp, nightsPurchased) {
  const entryMs = new Date(entryTimestamp).getTime();
  return new Date(entryMs + nightsPurchased * 24 * HOURS);
}

export function computeValidity(entryTimestamp, nightsPurchased, now = new Date()) {
  const expiry = computeExpiry(entryTimestamp, nightsPurchased);
  const nowMs = now instanceof Date ? now.getTime() : new Date(now).getTime();
  if (nowMs < expiry.getTime()) {
    return {
      expiry,
      status: "Active",
      remainingSeconds: Math.floor((expiry.getTime() - nowMs) / 1000),
    };
  }
  return { expiry, status: "Expired", remainingSeconds: 0 };
}

File: services/api/app/tickets.py
"""Ticket validity engine (build prompt Table 4).

    expiry = entry_timestamp + (nights_purchased x 24 hours)
    status = "Active" if now < expiry else "Expired"

Both the status and the remaining time are derived on every call and never
persisted, so a station with a wrong local clock cannot corrupt a stored expiry.
All arithmetic is done in timezone-aware UTC.
"""

import enum
from dataclasses import dataclass
from datetime import datetime, timedelta

from app.models.base import ensure_utc, utcnow


class TicketStatus(str, enum.Enum):
    ACTIVE = "Active"
    EXPIRED = "Expired"


@dataclass(frozen=True)
class TicketValidity:
    expiry: datetime
    status: TicketStatus
    remaining_seconds: int  # 0 once expired, never negative


def compute_expiry(entry_timestamp: datetime, nights_purchased: int) -> datetime:
    return ensure_utc(entry_timestamp) + timedelta(hours=24 * nights_purchased)


def compute_validity(
    entry_timestamp: datetime,
    nights_purchased: int,
    now: datetime | None = None,
) -> TicketValidity:
    reference = ensure_utc(now) if now is not None else utcnow()
    expiry = compute_expiry(entry_timestamp, nights_purchased)
    if reference < expiry:
        remaining = int((expiry - reference).total_seconds())
        return TicketValidity(expiry=expiry, status=TicketStatus.ACTIVE, remaining_seconds=remaining)
    return TicketValidity(expiry=expiry, status=TicketStatus.EXPIRED, remaining_seconds=0)
